client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +28,388 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::OVERLOADED)):
+        LOG_INFO(logger, "Request failed with OVERLOADED error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "All clients occupied");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::UNKNOWN)):
+        LOG_INFO(logger, "Request failed with UNKNOWN error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Failure due to unknown reason");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::INTERNAL_ERROR)):
+        LOG_INFO(logger, "Request failed with INTERNAL error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Failure due to some internal error");
+        break;
+      default:
+        LOG_INFO(logger, "Request failed with INTERNAL error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Internal error");
+        break;
+    }
+    return;
+  }
+  return_status = grpc::Status::OK;
+  auto reply = std::get<bft::client::Reply>(send_result);
+  if (is_rsi) {
+    ReconfigurationResponse res;
+    concord::messages::deserialize(reply.matched_data, res);
+    response.response.template emplace<ResponseT>(std::get<ResponseT>(res.response));
+  } else {
+    std::vector<std::unique_ptr<ResponseT>> responses;
+    for (const auto& r : reply.rsi) {
+      ReconfigurationResponse res;
+      concord::messages::deserialize(r.second, res);
+      responses.emplace_back(std::make_unique<ResponseT>(std::get<ResponseT>(res.response)));
+    }
+    response.response.template emplace<std::vector<std::unique_ptr<ResponseT>>>(std::move(responses));
+  }
+}
+
 Status StateSnapshotServiceImpl::GetRecentSnapshot(ServerContext* context,
                                                    const GetRecentSnapshotRequest* proto_request,
                                                    GetRecentSnapshotResponse* response) {
-  // TODO: Add implementation
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "GetRecentSnapshot");
+  std::chrono::milliseconds timeout = 5s;  // This is default timeout
+  if (context != nullptr) {
+    auto time_left =
+        gpr_time_to_millis(gpr_time_sub(gpr_now(context->raw_deadline().clock_type), context->raw_deadline()));
+    std::chrono::milliseconds curr_timeout{time_left};
+    if (curr_timeout.count() > 0) {
+      timeout = curr_timeout;
+    }
+  }
+
+  WriteConfig write_config{RequestConfig{false, 0, 1024 * 1024, timeout, "snapshotreq", "", false, true},
+                           bft::client::LinearizableQuorum{}};
+  StateSnapshotResponse snapshot_response;
+  grpc::Status return_status;
+  auto callback =
+      [&snapshot_response, &return_status, &write_config](concord::client::concordclient::SendResult&& send_result) {
```

Comment:
thanks, @robem, yes this is a valid concern. Since we are always sending a reconfiguration request we are not doing any client batching. So for our case client batching is always false. So we select the client and send the request then and there. But in the future reconfiguration requests might also be batched. So as you suggested, I will change the callback to async mode and receive the response in the callback, in the next check-in. Thanks for your meticulous review and comments.